sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala

```diff
@@ -2029,4 +2029,13 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
       testData2.select(lit(7), 'a, 'b).orderBy(lit(1), lit(2), lit(3)),
       Seq(Row(7, 1, 1), Row(7, 1, 2), Row(7, 2, 1), Row(7, 2, 2), Row(7, 3, 1), Row(7, 3, 2)))
   }
+
+  test("SPARK-21746: nondeterministic expressions correctly for filter predicates") {
+    withTempPath { path =>
+      val p = path.getAbsolutePath
+      Seq(1 -> "a").toDF("a", "b").write.partitionBy("a").parquet(p)
+      val df = spark.read.parquet(p)
+      checkAnswer(df.filter(rand(10) <= 1.0).select($"a"), Row(1))
```

Comment:
why this test case will trigger `InterpretedPredicate`?

Reply:
I ran this test without any exception. Are you sure this test can reproduce this issue?

Reply:
First, you don't add any comment to explain it. Reviewers don't know that.
Second, a new test should test against some regressions or the proposed changes. This test is neither of both.

Reply:
The test you add should trigger `InterpretedPredicate` as @cloud-fan mentioned. It should hit the code path you modified in this PR, not depending on another PR.

Reply:
sorry，I update the description of PR. thanks.

Reply:
can this test expose the bug?

Reply:
I'm updating to the latest version for validation.

Reply:
@viirya, @cloud-fan 
This trigger condition is associated with #18918. It will be more prone to this exception
The current spark master branch does not trigger this code path.
put this change on #18918 and close this PR.
thanks.